include/benchmark/benchmark_api.h

```diff
@@ -227,6 +234,113 @@ inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value) {
+
+  std::string const& Name() const { return name_; }
+  double Value() const { return value_; }
+
+  void Set(double v) { value_ = v; }
+
+  Counter& operator+= (double v) { value_ += v; return *this; }
+  Counter& operator-= (double v) { value_ -= v; return *this; }
+  Counter& operator*= (double v) { value_ *= v; return *this; }
+  Counter& operator/= (double v) { value_ /= v; return *this; }
+
+  void Finish(double cpu_time, double num_threads);
+
+private:
+
+  std::string name_;
+  double      value_;
+  uint32_t    flags_;
+  // <jppm> maybe we could add a description field...
+
+};
+
+
+
+class BenchmarkCounters {
+public:
+
+  typedef Counter                         value_type;
+  typedef std::vector< Counter >          container_type;
+  typedef container_type::const_iterator  const_iterator;
+
+public:
+
+  BenchmarkCounters(size_t initial_capacity = 16);
+
+  // Add/Set a counter. Will overwrite any additional counter properties.
+  size_t  Set(Counter const& c);
+  // Add/Set a counter. When the counter already exists, f is ignored.
+  size_t  Set(const char*        n, double v, uint32_t f = Counter::kDefaults);
+  // Add/Set a counter. When the counter already exists, f is ignored.
+  size_t  Set(std::string const& n, double v, uint32_t f = Counter::kDefaults);
+
+#ifdef BENCHMARK_INITLIST
+  void    Set(std::initializer_list< Counter > il) { for(auto &c : il) { Set(c); } }
+#endif
+
+  // Returns the index where name is located, or size() if the name was not found.
+  size_t  Find(const char*        name) const;
+  // Returns the index where name is located, or size() if the name was not found.
+  size_t  Find(std::string const& name) const;
+
+  bool    Exists(const char*        name) const { return Find(name) != counters_.size(); }
+  bool    Exists(std::string const& name) const { return Find(name) != counters_.size(); }
+
+  Counter      & Get(size_t               id);
+  Counter const& Get(size_t               id) const;
+  Counter      & Get(const char*        name);
+  Counter const& Get(const char*        name) const;
+  Counter      & Get(std::string const& name);
+  Counter const& Get(std::string const& name) const;
+
+  bool    SameNames(BenchmarkCounters const& that) const;
+
+  void    Sum(BenchmarkCounters const& that);
+  void    Finish(double time, double num_threads);
+  void    Clear();
+
+  size_t          size() const { return counters_.size();  }
+  const_iterator begin() const { return counters_.begin(); }
+  const_iterator end  () const { return counters_.end();   }
+
+/* <jppm> Keeping for now the old bytes_processed and items_processed.
+public:
+
+  // Common counters
+  Counter const& GetBytesPerSecond() const;
+  Counter const& GetItemsPerSecond() const;
+  size_t  BytesPerSecond() const;
+  size_t  ItemsPerSecond() const;
+  void    BytesPerSecond(size_t b);
+  void    ItemsPerSecond(size_t i);
+*/
+
+private:
+
+  std::vector< Counter > counters_;
```

Comment:
std::map<const char*, Counter> might be better. You can avoid the 'find' cost on every API access.